File: today/reverse-linked-list-ii.ts
/**
 * https://leetcode.com/problems/reverse-linked-list-ii/description/
 */
export { reverseBetween, toList, toArray };

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
  return null;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function toList(elements: number[]): ListNode | null {
  if (elements.length === 0) return null;
  let prev: ListNode | null = null;
  for (let i = elements.length - 1; i >= 0; i--) {
    const node = new ListNode(elements[i]);
    node.next = prev;
    prev = node;
  }
  return prev;
}

function toArray(list: ListNode | null): number[] {
  if (!list) return [];
  const result: number[] = [];
  while (list) {
    result.push(list.val);
    list = list.next;
  }
  return result;
}

File: returned/python/today7/reverse_linked_list_ii.py
# https://leetcode.com/problems/reverse-linked-list-ii/description/
from typing import List, Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(None, head)
        pointer = dummy
        before_nodes = None        
        nodes: List[ListNode] = []

        count = 0
        while count <= right:
            if count == left - 1:
                before_nodes = pointer
            if count >= left:
                nodes.append(pointer)
            count += 1
            pointer = pointer.next

        after_nodes = nodes[-1].next

        for i in range(len(nodes)-1, 0, -1):
            nodes[i].next = nodes[i-1]
        
        nodes[0].next = after_nodes
        before_nodes.next = nodes[-1]

        return dummy.next